### ISSUE TITLE
asset_get_all(asset_type) - A way of retrieving all assets in an array by type

### ISSUE THREAD

**Issue by tabularelf** [YoYoGames/GameMaker-Bugs#3039]
### Is your feature request related to a problem?

Normally this wouldn't make sense to add in, as assets are just numbers in the current runtime, and this would only matter for the new runtime, since they will be asset refs instead of asset ids by that point. **However as we speak**, this is happening **right now** in the current beta for the current runtime. And to be honest, having such a change without at least having some way of fetching all assets of a given type, is a very poor move.

Libraries like Scribble by JujuAdams for instance, indexes and adds all of the games fonts to be used with Scribble. I myself, have Sonus, which indexes and categories all sounds into two OOP-like constructors based on if they're streamed or in memory. (Sonus supports in adding in external files as audio). And libraries like YUI by shdwcat, and GMLive by YellowAfterLife, both rely on searching up the entire function dictionary to be added for the users benefit.

Just these libraries alone for instance, will break. And there isn't strictly an easier way of fixing these, as these are meant to be plug and play. I also have an entire game that has a whole modding implementation for coding as well. And adding in a huge list of functions (estimating anywhere from 200-400+ at this given time, but could easily be more overtime) manually isn't strictly my cup of tea. Let alone, it also destroys my `unsafe_get` function (that though, also does include built in function). Both of which absolutely do require a database for quick accessing fetching function names. 

This is something I can't easily do, and would require a good chunk of my time indexing each function. It's arguably a lot easier adding in a blacklist than it is adding in a whitelist. Especially for someone who wants to add in a bunch of library functions, and then just remove the ones I don't want. i.e.
```gml
ModAPI.AssignFunctionByName("scribble"); // This only adds functions that start with "scribble". So scribble and scribble_* functions
ModAPI.RemoveFunction("scribble_msdf_thickness_offset", "scribble_flush_everything", "scribble_external_sound_add", "scribble_external_sound_exists", "scribble_external_sound_remove");
```

### Describe the solution you'd like

The idea I have in mind is that we can do `asset_get_all(asset_type)`, where `asset_type` will be one of the many asset keywords. Such as:
- `asset_object`
- `asset_sprite`
- `asset_sound`
- `asset_room`
- `asset_tiles`
- `asset_path`
- `asset_script` 
- `asset_font`
- `asset_timeline`
- `asset_shader`
- `asset_animationcurve`
- `asset_sequence`
- `asset_particlesystem`

(`asset_script` should refer to functions like `function foo() {}` imo, or introduce `asset_function`).
This also covers assets that were added dynamically on the fly. The return value will be an array of all of the assets, specified by the asset type.
The benefit of this approach is that for a developer and for modding purposes, it makes it a lot easier to reference and pull things directly later on for console usage, a custom programming language or having an overall lookup of every asset needed. It also does allow us library makers to at least update and adapt the new "asset ids are now refs" change, without much of a hitch overall.
Giving more advanced users, and library makers if I didn't make it clear enough, a way of referencing assets properly.
```gml
/// Script: game_init
global.spriteMap = {};
var _sprites = asset_get_all(asset_sprite);
for(var _i = 0, _len = array_length(_sprites); _i < _len; ++_i) {
    var _sprite = _sprites[_i];
    var _name = sprite_get_name(_sprite);
    global.spriteMap[$ _name] = _sprite;
}
```

### Describe alternatives you've considered

The change for this does make sense for asset ids to become asset refs, but as it stands, the only other alternatives are either:
- Tell YYG not to do this (which imo, isn't great for either of us)
- Me, the end user, stay on 2023.4 for a very long time. (Which is quite sour imo too, as I'm stuck on a much older version for longer when I could be on the latest monthly. Bugs aside)
- Telling every other user who wants to use Scribble, Sonus, GMLive, YUI or any other library that requires going through a list of asset indexes on boot, to downgrade their GameMaker version to before asset ids were changed. (Which imo, also doesn't look great for either YYG or the library itself.)
- Manually adding in every. Single. Asset to wherever it's needed. (Tags would be great here... Except that there is no **autotagger** for libraries as it stands. And deleting/reimporting libraries for updates also removes said tags. So this is pretty stupid)

### Additional context

The main big reminder is that, asset ids are refs, is currently in beta right now. Meaning that if this isn't added in by the time this hits monthly, this can and will break for a lot of people. As it'll affect libraries that do add assets straight away for the developers convenience. This would require putting in *more work* for not just the library developer that does need this, but for the users who do want to use said library too.

**Comment by tabularelf** [comment]
We've got `asset_get_ids`, which fulfil the above request.

**Comment by YYDan** [comment]
Yes, this should have been closed as a duplicate a long time ago ;)

See https://github.com/YoYoGames/GameMaker-Bugs/issues/2964 for the info on that function above, which was released in 2023.11.
